src/components/utils/segment.ts

```diff
@@ -36,6 +42,104 @@ export function segmentHasOrderBy(
   );
 }
 
+// Collects a mapping from names in the output space to
+// the full name of the fields that are associated with the
+// output names, if that mapping could be computed.
+function getOuputNameToInputNameMap(
+  segment: ASTSegmentViewDefinition
+): Map<string, string> {
+  const nameMap = new Map<string, string>();
+
+  for (const operation of segment.operations.items) {
+    if (operation instanceof ASTGroupByViewOperation) {
+      if (operation.node.field.expression.kind === 'field_reference') {
+        const expression = operation.node.field.expression;
+        nameMap.set(
+          operation.name || expression.name,
+          toFullName(expression.path, expression.name)
+        );
+      }
+    } else if (
+      operation instanceof ASTAggregateViewOperation &&
+      operation.name
+    ) {
+      if (operation.node.field.expression.kind === 'field_reference') {
+        const expression = operation.node.field.expression;
+        nameMap.set(
+          operation.name || expression.name,
+          toFullName(expression.path, expression.name)
+        );
+      }
+    }
+  }
```

Comment:
The query builder accessors can do a lot of lifting here. `getReference()` is the method Chris mentioned, and `operation.name` is a getter that is either the operator name or the expression name so covers that base.

```suggestion
  for (const operation of segment.operations.items) {
    if (
      operation instanceof ASTGroupByViewOperation ||
      operation instanceof ASTAggregateViewOperation
    ) {
      const {name, path} = operation.field.getReference();
      nameMap.set(operation.name, toFullName(path, name));
    }
  }
```

Reply:
That does help reduce some of the duplication - thanks!